tensorpipe/test/channel/channel_test.cc

```diff
@@ -15,513 +15,568 @@
+    std::shared_ptr<Context<TTensor>> ctx =
+        this->helper_->makeContext("client");
+    auto channel = ctx->createChannel(std::move(conn), Endpoint::kConnect);
+
+    // Initialize with sequential values.
+    std::vector<uint8_t> data(dataSize);
+    std::iota(data.begin(), data.end(), 0);
+    DataWrapper<TTensor> wrappedData(data);
+
+    // Perform send and wait for completion.
+    std::future<std::tuple<Error, TDescriptor>> descriptorFuture;
+    std::future<Error> sendFuture;
+    std::tie(descriptorFuture, sendFuture) =
+        sendWithFuture(channel, wrappedData.tensor());
+    Error descriptorError;
+    TDescriptor descriptor;
+    std::tie(descriptorError, descriptor) = descriptorFuture.get();
+    EXPECT_FALSE(descriptorError) << descriptorError.what();
+    this->peers_->send(PeerGroup::kServer, descriptor);
+    Error sendError = sendFuture.get();
+    EXPECT_FALSE(sendError) << sendError.what();
+
+    this->peers_->done(PeerGroup::kClient);
+    this->peers_->join(PeerGroup::kClient);
+
+    ctx->join();
+  }
+};
+
+CHANNEL_TEST_GENERIC(ServerToClient);
+
+template <typename TTensor>
+class SendMultipleTensorsTest : public ChannelTest<TTensor> {
+  const int dataSize = 256 * 1024; // 256KB
+  static constexpr int numTensors = 100;
+
+ public:
+  void server(std::shared_ptr<transport::Connection> conn) override {
+    std::shared_ptr<Context<TTensor>> ctx =
+        this->helper_->makeContext("server");
+    auto channel = ctx->createChannel(std::move(conn), Endpoint::kListen);
+
+    // Initialize with sequential values.
+    std::vector<uint8_t> data(dataSize);
+    std::iota(data.begin(), data.end(), 0);
+    DataWrapper<TTensor> wrappedData(data);
+
+    // Error futures
+    std::vector<std::future<Error>> sendFutures;
+
+    // Perform send and wait for completion.
+    for (int i = 0; i < numTensors; i++) {
+      std::future<std::tuple<Error, TDescriptor>> descriptorFuture;
+      std::future<Error> sendFuture;
+      std::tie(descriptorFuture, sendFuture) =
+          sendWithFuture(channel, wrappedData.tensor());
+      Error descriptorError;
+      TDescriptor descriptor;
+      std::tie(descriptorError, descriptor) = descriptorFuture.get();
+      EXPECT_FALSE(descriptorError) << descriptorError.what();
+      this->peers_->send(PeerGroup::kClient, descriptor);
+      sendFutures.push_back(std::move(sendFuture));
+    }
+    for (auto& sendFuture : sendFutures) {
+      Error sendError = sendFuture.get();
+      EXPECT_FALSE(sendError) << sendError.what();
+    }
+
+    this->peers_->done(PeerGroup::kServer);
+    this->peers_->join(PeerGroup::kServer);
+
+    ctx->join();
+  }
+
+  void client(std::shared_ptr<transport::Connection> conn) override {
+    std::shared_ptr<Context<TTensor>> ctx =
+        this->helper_->makeContext("client");
+    auto channel = ctx->createChannel(std::move(conn), Endpoint::kConnect);
+
+    std::vector<std::unique_ptr<DataWrapper<TTensor>>> wrappedDataVec;
```

Comment:
Do we need to wrap the DataWrappers in unique_ptrs because they are not movable? Would it make sense perhaps to make them movable? As it stands, we're making plenty of heap allocations just so they can hold a few pointers...

Reply:
Yes, making them movable is probably the right thing to do here. Will update.